haystack/nodes/label_generator/pseudo_label_generator.py

```diff
@@ -0,0 +1,211 @@
+import sys
+import random
+from typing import List, Dict, Union, Tuple
+
+from sentence_transformers import CrossEncoder
+
+from haystack.nodes import BaseRetriever, QuestionGenerator
+from haystack.schema import Document
+from haystack.nodes.base import BaseComponent
+
+
+class PseudoLabelGenerator(BaseComponent):
+    """
+    The PseudoLabelGenerator is a component that creates Generative Pseudo Labeling (GPL) training data for training
+    of dense retrievers.
+
+    GPL is an unsupervised domain adaptation method for training of dense retrievers. It is based on question generation
+    and pseudo labeling with powerful cross-encoders. To train a domain-adapted model, it needs access to unlabeled
+    target corpus usually via DocumentStore and a retriever to mine for negatives.
+
+    For more details see https://github.com/UKPLab/gpl
+
+    For example:
+
+    ```python
+    |   document_store = DocumentStore(...)
+    |   retriever = Retriever(...)
+    |   qg = QuestionGenerator(model_name_or_path="doc2query/msmarco-t5-base-v1")
+    |   plg = PseudoLabelGenerator(qg, retriever)
+    |   train_examples = []
+    |   for idx, doc in enumerate(document_store):
```

Comment:
I think we wouldn't need to iterate here over each document individually. We could also just use `document_store.get_all_documents()` and provide this as input.